tests/test_coordinate_source.py

```diff
@@ -0,0 +1,328 @@
+        imgs = self.src_from_box.images(0, 440)
+        # make sure max_rows loads the correct particles
+        src_100 = EmanCoordinateSource(self.files_box, max_rows=100)
+        imgs_100 = src_100.images(0, src_100.n)
+        for i in range(100):
+            self.assertTrue(np.array_equal(imgs[i], imgs_100[i]))
+        # make sure max_rows > self.n loads max_rows images
+        src_500 = EmanCoordinateSource(self.files_box, max_rows=500)
+        self.assertEqual(src_500.n, 440)
+        imgs_500 = src_500.images(0, 440)
+        for i in range(440):
+            self.assertTrue(np.array_equal(imgs[i], imgs_500[i]))
+        # make sure max_rows loads correct particles
+        # when some have been excluded
+        imgs_newsize = EmanCoordinateSource(self.files_box, particle_size=336).images(
+            0, 50
+        )
+        src_maxrows = EmanCoordinateSource(
+            self.files_box, particle_size=336, max_rows=50
+        )
+        # max_rows still loads 50 images even if some particles were excluded
+        self.assertEqual(src_maxrows.n, 50)
+        imgs_maxrows = src_maxrows.images(0, 50)
+        for i in range(50):
+            self.assertTrue(np.array_equal(imgs_newsize[i], imgs_maxrows[i]))
+
+    def testBoundaryParticlesRemoved(self):
+        src_centers_larger_particles = CentersCoordinateSource(
+            self.files_coord, particle_size=300
+        )
+        src_box_larger_particles = EmanCoordinateSource(
+            self.files_box, particle_size=300
+        )
+        # 2 particles do not fit at this particle size
+        self.assertEqual(src_centers_larger_particles.n, 438)
+        self.assertEqual(src_box_larger_particles.n, 438)
+        # make sure we have the same particles
+        imgs_centers = src_centers_larger_particles.images(0, 438)
+        imgs_resized = src_box_larger_particles.images(0, 438)
+        for i in range(50):
+            self.assertTrue(np.array_equal(imgs_centers[i], imgs_resized[i]))
+
+    def testEvenOddResize(self):
+        # test a range of even and odd resizes
+        for _size in range(252, 260):
+            src_centers = CentersCoordinateSource(self.files_coord, particle_size=_size)
+            src_resized = EmanCoordinateSource(self.files_box, particle_size=_size)
+            imgs_centers = src_centers.images(0, 440)
+            imgs_resized = src_resized.images(0, 440)
+            for i in range(440):
+                self.assertTrue(np.array_equal(imgs_centers[i], imgs_resized[i]))
+
+    def testSave(self):
+        # we can save the source into an .mrcs stack with *no* metadata
+        src = EmanCoordinateSource(self.files_box, max_rows=10)
+        imgs = src.images(0, 10)
+        star_path = os.path.join(self.tmpdir.name, "stack.star")
+        mrcs_path = os.path.join(self.tmpdir.name, "stack_0_9.mrcs")
+        src.save(star_path)
+        # load saved particle stack
+        saved_mrcs_stack = mrcfile.open(mrcs_path).data
```

Comment:
How do we know that the data is stored in `stack_0_9.mrcs`?

Reply:
Behavior of `ImageSource.save()`:

```
fstem = os.path.splitext(fname)[0]
mrcs_filename = f"{fstem}_{0}_{self.n-1}.mrcs"
```

This will always save as the STAR file basename plus the range of indices being saved (even in single save mode)

Reply:
Eh that's somewhat brittle. Not sure what a good solution is here. One would be to have `src.save` return the name of the mrc file that it used. The other would be to read in the STAR file and look at what the MRC file is. I have some memory of discussing this a while back (when Junchao was implementing this) but I don't recall what the outcome was.

Reply:
So modify `ImageSource.save()` as part of this PR? There would then be an issue of what to return when save mode is "multiple". A list?

Reply:
I agree that returning the filename(s) is useful.

Reply:
> So modify `ImageSource.save()` as part of this PR? There would then be an issue of what to return when save mode is "multiple". A list?
> I agree that returning the filename(s) is useful.

No that would better to do at some later point. Returning a list of files sounds reasonable in the multiple case, but it's hard to predict how the interface will evolve. Let's make an issue. For now, would it be possible to read in the STAR file and get the MRC file there?

Reply:
Issue #545 

I can get the path from the STAR file but it's a bit ugly